File: src/main/java/unaverage/strategic_ench/config/BaneOfArthropod.java
package unaverage.strategic_ench.config;

import com.google.common.collect.ImmutableSet;
import com.typesafe.config.Config;
import net.minecraft.entity.EntityType;
import net.minecraft.util.registry.Registry;
import unaverage.strategic_ench.Helper;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * Contains configs relating to change to the bane of arthropod enchantment
 */
public class BaneOfArthropod {
    static final Map<String, Object> DEFAULT = Map.of(
        ExtraAffectedMobs.class.getSimpleName(), ExtraAffectedMobs.DEFAULT
    );

    final ExtraAffectedMobs extraAffectedMobs;

    BaneOfArthropod(Config c) {
        c = c.getConfig(this.getClass().getSimpleName());

        extraAffectedMobs = new ExtraAffectedMobs(c);
    }

    /**
     * {@return true if the entity is affected by bane of arthropods and was not affected in vanilla, false otherwise}
     * @param e The {@link EntityType} of the entity being tested
     */
    public boolean isExtraAffectedMob(EntityType<?> e){
        return Helper.contains(
            e,
            extraAffectedMobs.value,
            Registry.ENTITY_TYPE
        );
    }

    static class ExtraAffectedMobs {
        static final Set<String> DEFAULT = ImmutableSet.of(
            "minecraft:guardian",
            "minecraft:elder_guardian"
        );

        public Set<String> value;

        ExtraAffectedMobs(Config c) {
            value = new HashSet<>(
                c.getStringList(this.getClass().getSimpleName())
            );
        }
    }

    //TODO
    static class BreakCobWebFaster{
        static final boolean DEFAULT = true;

        final boolean value;

        BreakCobWebFaster(Config c){
            value = c.getBoolean(this.getClass().getSimpleName());
        }
    }

    //TODO
    static class LessCobWebSlowDown{
        static final boolean DEFAULT = true;

        final boolean value;

        LessCobWebSlowDown(Config c){
            value = c.getBoolean(this.getClass().getSimpleName());
        }
    }
}

File: src/main/kotlin/unaverage/strategic_ench/config/BaneOfArthropod.kt
package unaverage.strategic_ench.config

import com.google.common.collect.ImmutableSet
import net.minecraft.entity.EntityType
import unaverage.strategic_ench.config.BaneOfArthropod.ExtraAffectedMobs

/**
 * Contains configs relating to change to the bane of arthropod enchantment
 */
class BaneOfArthropod internal constructor() {
    val extraAffectedMobs: ExtraAffectedMobs

    init {
        extraAffectedMobs = ExtraAffectedMobs()
    }

    /**
     * {@return true if the entity is affected by bane of arthropods and was not affected in vanilla, false otherwise}
     * @param e The [EntityType] of the entity being tested
     */
    fun isExtraAffectedMob(e: EntityType<*>?): Boolean {
        return false
    }

    class ExtraAffectedMobs {
        var value: Set<String>

        init {
            value = HashSet()
        }

        companion object {
            fun getDefault(newFile: Boolean): Set<String> {
                return if (!newFile) java.util.Set.of() else ImmutableSet.of(
                    "minecraft:guardian",
                    "minecraft:elder_guardian"
                )
            }
        }
    }

    companion object {
        fun getDefault(newFile: Boolean): Map<String, Any> {
            return java.util.Map.of<String, Any>(
                ExtraAffectedMobs::class.java.simpleName, ExtraAffectedMobs.getDefault(newFile)
            )
        }
    }
}